src/preamble.js

```diff
@@ -246,6 +242,13 @@ function initRuntime() {
   callRuntimeCallbacks(__RELOC_FUNCS__);
 #endif
   <<< ATINITS >>>
+  
+#if WASM_WORKERS
+if (ENVIRONMENT_IS_WASM_WORKER) {
+  return _wasmWorkerInitializeRuntime();
+}
+#endif
```

Comment:
This change does not look correct. `_wasmWorkerInitializeRuntime()` initializes the stack and TLS of the Worker, which is to be done instead of `checkStackCookie();` and `setStackLimits();` above.

It loks like after this code, the stack would be initialized twice, and `__RELOC_FUNCS__` and `<<< ATINITS >>>` would also be run in Wasm Workers, but they aren't run in Pthreads either?

What was the motivation for this change?

Reply:
Since `_wasmWorkerInitializeRuntime()` returns early before `<<< ATINITS >>>`, cause `FS.init` in <<< ATINITS >>> is not been called, resulting in stdin(0), stdout(1) and stderr(2) these three device is not been created in WASM Worker environment.
When the code uses pipes for output (such as printf or cout), will throw a device dose not exists error. Especially when the AddressSanitizer tool is enabled by `-fsanitize=address` flag.
So I modified the order to ensure `<<< ATINITS >>>` can be executed.

Reply:
I adjusted the calling order so that now only the methods in `_wasmWorkerInitializeRuntime` and `<<< ATINITS >>>` will be called in WASM Worker environment.

Reply:
The semantics of <<< ATINITS >>> is to not be called in startup of pthreads either (note the if (ENVIRONMENT_IS_PTHREAD) return; block), so the semantics should equally be preserved to not be called at startup of wasm workers as well.

Some people might have ATINITS that should only run once at whole program startup, so this would change the semantics on them that they'd be inadvertently rerunning whenever a new wasm worker is created, which would break their programs.

I think we may need a custom solution for the wasm worker stdio situation, maybe something that parallels pthreads. I forget, how did pthreads get to initialize the stdio handles so printing works there?

Reply:
IIUC, with pthreads all FS calls are proxied back to the main thread.  i.e. None of FS API is designed to be called from a worker.

Reply:
Oh right..

I wonder if using WASMFS avoids all of that to make it work without hassle?

Maybe we'd want to do a separate `<<< WORKER_ATINITS >>>` type of section that would run for each Worker startup?

There library_fs.js could run the needed stdio setup code that is desirable for each Worker context, and in this case gate it behind `if (ENVIRONMENT_IS_WASM_WORKER)`. (That kind of mechanism might be more generic than crafting a Wasm Worker specific <<< WASM_WORKER_ATINITS >>> block)

Reply:
Are the filesystem APIs expected to work in wasm worker though?    I don't really see how they could.  Shouldn't try assert/abort because proxying is not available?

Reply:
Hey, guys,
I found an alternative solutions to solve this problem. 
Instead of modifying the init script of the WASM worker, I adjusted the implementation of `fd_write(...)` method in `library_wasi.js`.
When the code runs inside the WASM worker, use a simple way to support stdout and stderr pipes.